### PR TITLE
Make ‘Document.find’ return an instance of the right subclass

### PR DESCRIPTION
**Pull request by tuzz**
This makes `Document.find` return a `CmaCase` if the `document_type` is `cma_case`, for example.

### CODE REVIEW

File: app/models/document.rb [hunk 1]
```diff
@@ -184,14 +184,22 @@ def self.all(page, per_page, q: nil)
...
+    document_type = attributes.fetch("document_type")
+    document_class = document_type.camelize.constantize
+
+    if [document_class, Document].include?(self)
+      document_class.from_publishing_api(response.to_hash)
     else
-      raise RecordNotFound
+      message = "#{self}.find('#{content_id}') returned the wrong type: '#{document_class}'"
+      raise TypeMismatchError, message
     end
```
Comment (h-lame) [inline app/models/document.rb:198]: I find this conditional a bit hard to parse, but I also don't have any better idea of how to express it so :shrugging_emoji_on_loop: ?

**Comment by h-lame** [comment]
Unless anyone has a better idea about the conditional I mentioned above, I reckon this is good to merge - makes things way less surprising.

File: app/models/document.rb (hunk 1 above)
Comment (tuzz) [inline app/models/document.rb:198]: Would either of these be any clearer?

1) 

``` ruby
if document_class == self || Document == self
```

2) 

``` ruby
if document_class == self
  # ...
elsif Document == self
  # ...
```

Comment (h-lame) [inline app/models/document.rb:198]:
``` ruby
if document_class == self || self.descendants.include? document_class
  #...
else
  #...
end
```

Maybe?  Not sure how reliable `descendants` would be - but then we do constantize before we do anything so it should be ok?

Comment (h-lame) [inline app/models/document.rb:198]: TBH I reckon I'm dangerously close to having an opinion on the colour of this bikeshed - most likely it's fine as-is.

Comment (tuzz) [inline app/models/document.rb:198]: I think `descendants` might be incorrect if Rails autoloading hasn't loaded all of the models yet. I'm going to merge as-is and we can change it later if needed.
